File: wrappers/dotnet/indy-sdk-dotnet-test/AnonCredsTests/IssuerRevokeClaimTest.cs
using Hyperledger.Indy.AnonCredsApi;
using Hyperledger.Indy.BlobStorageApi;
using Hyperledger.Indy.Test.Util;
using Hyperledger.Indy.WalletApi;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Newtonsoft.Json.Linq;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.AnonCredsTests
{
    [TestClass]
    public class IssuerRevokeClaimTest : AnonCredsIntegrationTestBase
    {
        private const string WALLET_NAME = "commonWallet";
        private const string WALLET_KEY = "commonWalletKey";
        private Wallet _issuerWallet;
        private const string _walletName = "issuerWallet";
        private string _tailsWriterConfig = string.Format("{{\"base_dir\":\"{0}\", \"uri_pattern\":\"\"}}", "TODO");
        private const int _userRevocIndex = 1;
        private const string _proofReqJson = "{" +
                                       "\"nonce\":\"123432421212\"," +
                                       "\"name\":\"proof_req_1\"," +
                                       "\"version\":\"0.1\"," +
                                       "\"requested_attrs\":{\"attr1_referent\":{\"name\":\"name\",\"restrictions\":[{\"schema_seq_no\":1}]}}," +
                                       "\"requested_predicates\":{}" +
                                    "}";    
        
        private const string _requestedClaimsJsonTemplate = "{{" +
                                                      "\"self_attested_attributes\":{{}}," +
                                                      "\"requested_attrs\":{{\"attr1_referent\":[\"{0}\", true]}}," +
                                                      "\"requested_predicates\":{{}}" +
                                                    "}}";
        private string _claimDefJson;
        private IssuerCreateCredentialResult _claimResult;

        [TestInitialize]
        public async Task Before()
        {
            StorageUtils.CleanupStorage();

            //1. Create Issuer wallet, get wallet handle
            await WalletUtils.CreateWallet(WALLET_NAME, WALLET_KEY);
            _issuerWallet = await WalletUtils.OpenWallet(WALLET_NAME, WALLET_KEY);

            //2. Issuer create claim definition
            IssuerCreateAndStoreCredentialDefResult result = await AnonCreds.IssuerCreateAndStoreCredentialDefAsync(_issuerWallet, issuerDid, schema, null, null, null);
            _claimDefJson = result.CredDefJson;

            //3. Issuer create revocation registry
            BlobStorageWriter tailsWriter = await BlobStorage.OpenWriterAsync("default", _tailsWriterConfig);
            await AnonCreds.IssuerCreateAndStoreRevocRegAsync(_issuerWallet, issuerDid, null, null, null, null, tailsWriter);

            //4. Prover create Master Secret
            await AnonCreds.ProverCreateMasterSecretAsync(_issuerWallet, masterSecretName);

            //5. Prover store Claim Offer received from Issuer
            var claimOfferJson = string.Format(claimOfferTemplate, issuerDid, 1);
            // TODO await AnonCreds.ProverStoreCredentialOfferAsync(_issuerWallet, claimOfferJson);

            //6. Prover create Claim Request
            var proverDid = "BzfFCYk";
            var claimReq = await AnonCreds.ProverCreateCredentialReqAsync(
                _issuerWallet,
                proverDid,
                claimOfferJson,
                _claimDefJson,
                masterSecretName);

            //7. Issuer create Claim
            var claimJson = "{" +
                "\"sex\":[\"male\",\"5944657099558967239210949258394887428692050081607692519917050011144233115103\"]," +
               "\"name\":[\"Alex\",\"1139481716457488690172217916278103335\"]," +
               "\"height\":[\"175\",\"175\"]," +
               "\"age\":[\"28\",\"28\"]" +
            "}";

            
            _claimResult = null; // TODO await AnonCreds.IssuerCreateCredentialAsync(_issuerWallet, claimReq, claimJson, _userRevocIndex);

            //8. Prover store received Claim
            // TODO await AnonCreds.ProverStoreClaimAsync(_issuerWallet, _claimResult.CredentialJson, _claimResult.RevocRegDeltaJson);
        }

        [TestCleanup]
        public async Task After()
        {
            await _issuerWallet.CloseAsync();
            StorageUtils.CleanupStorage();
        }

        [TestMethod]
        public async Task TestAnoncredsWorksForClaimRevokedBeforeProofCreated()
        {
            //9. Issuer revoke claim
            var revocRegUpdateJson = ""; // TODO await AnonCreds.IssuerRevokeCredentialAsync(
                //_issuerWallet,
                //issuerDid,
                //1,
                //1);

            //10. Prover gets Claims for Proof Request
            var claimsJson = await AnonCreds.ProverGetCredentialsForProofReqAsync(_issuerWallet, _proofReqJson);
            var claims = JObject.Parse(claimsJson);
            var claimsForAttr1 = claims["attrs"]["attr1_referent"];
            var claim = claimsForAttr1[0];
            var claimUuid = claim.Value<string>("referent");

            //11. Prover create Proof
            var requestedClaimsJson = string.Format(_requestedClaimsJsonTemplate, claimUuid);

            var schemasJson = string.Format("{{\"{0}\":{1}}}", claimUuid, schema);
            var claimDefsJson = string.Format("{{\"{0}\":{1}}}", claimUuid, _claimDefJson);
            var revocRegsJsons = string.Format("{{\"{0}\":{1}}}", claimUuid, revocRegUpdateJson);

            var ex = await Assert.ThrowsExceptionAsync<ClaimRevokedException>(() =>
                AnonCreds.ProverCreateProofAsync(
                    _issuerWallet,
                    _proofReqJson,
                    requestedClaimsJson,
                    schemasJson,
                    masterSecretName,
                    claimDefsJson,
                    revocRegsJsons)
            );
        }

        [TestMethod]
        public async Task TestAnoncredsWorksForClaimRevokedAfterProofCreated()
        {
            //9. Prover gets Claims for Proof Request
            var claimsJson = await AnonCreds.ProverGetCredentialsForProofReqAsync(_issuerWallet, _proofReqJson);
            var claims = JObject.Parse(claimsJson);
            var claimsForAttr1 = claims["attrs"]["attr1_referent"];
            var claim = claimsForAttr1[0];
            var claimUuid = claim.Value<string>("referent");

            //10. Prover create Proof
            var requestedClaimsJson = string.Format(_requestedClaimsJsonTemplate, claimUuid);

            var schemasJson = string.Format("{{\"{0}\":{1}}}", claimUuid, schema);
            var claimDefsJson = string.Format("{{\"{0}\":{1}}}", claimUuid, _claimDefJson);
            var revocRegsJsons = string.Format("{{\"{0}\":{1}}}", claimUuid, _claimResult.RevocRegDeltaJson);

            var proofJson = await AnonCreds.ProverCreateProofAsync(
                _issuerWallet,
                _proofReqJson,
                requestedClaimsJson,
                schemasJson,
                masterSecretName,
                claimDefsJson,
                revocRegsJsons);

            //11. Issuer revoke prover claim
            // TODO
            //var revocRegUpdateJson = await AnonCreds.IssuerRevokeCredentialAsync(
            //    _issuerWallet,
            //    issuerDid,
            //    1,
            //    1);

            ////12. Verifier verify proof
            //var updatedRevocRegsJsons = string.Format("{{\"{0}\":{1}}}", claimUuid, revocRegUpdateJson);

            var valid = false; // TODO await AnonCreds.VerifierVerifyProofAsync(
                //_proofReqJson,
                //proofJson,
                //schemasJson,
                //claimDefsJson,
                //updatedRevocRegsJsons);

            Assert.IsFalse(valid);
        }
    }
}

File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/anoncreds/IssuerRevokeClaimTest.java
package org.hyperledger.indy.sdk.anoncreds;

import static org.hyperledger.indy.sdk.utils.EnvironmentUtils.getIndyHomePath;
import static org.junit.Assert.*;


import blob_storage.BlobStorage;
import org.hyperledger.indy.sdk.wallet.Wallet;
import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.*;


public class IssuerRevokeClaimTest extends AnoncredsIntegrationTest {

	private String tailsWriterConfig = String.format("{\"base_dir\":\"%s\", \"uri_pattern\":\"\"}", getIndyHomePath("tails"));

	@Test
	public void testIssuerRevokeProofWorks() throws Exception {
		//1. Create wallet, get wallet handle
		String walletName = "revocationWallet";
		Wallet.createWallet("default", walletName, "default", null, null).get();
		Wallet wallet = Wallet.openWallet(walletName, null, null).get();

		//2. Issuer create Schema
		AnoncredsResults.IssuerCreateSchemaResult createSchemaResult = Anoncreds.issuerCreateSchema(issuerDid, gvtSchemaName, schemaVersion, gvtSchemaAttributes).get();
		String schemaJson = createSchemaResult.getSchemaJson();

		//3. Issuer create claim definition
		String revocationCredentialDefConfig = "{\"support_revocation\":true}";
		AnoncredsResults.IssuerCreateAndStoreClaimDefResult createClaimDefResult = Anoncreds.issuerCreateAndStoreClaimDef(wallet, issuerDid, schemaJson, tag, null, revocationCredentialDefConfig).get();
		String claimDefId = createClaimDefResult.getClaimDefId();
		String claimDefJson = createClaimDefResult.getClaimDefJson();

		//4. Issuer create revocation registry
		String revRegConfig = "{\"issuance_type\":null,\"max_cred_num\":5}";
		AnoncredsResults.IssuerCreateAndStoreRevocRegResult createRevRegResult = Anoncreds.issuerCreateAndStoreRevocReg(wallet, issuerDid, null, tag, claimDefId, revRegConfig, "default", tailsWriterConfig).get();
		String revRegId = createRevRegResult.getRevRegId();
		String revRegDef = createRevRegResult.getRevRegDefJson();

		//5. Prover create Master Secret
		Anoncreds.proverCreateMasterSecret(wallet, masterSecretName).get();

		//6. Issuer create Claim Offer
		String claimOfferJson = Anoncreds.issuerCreateClaimOffer(wallet, claimDefId, issuerDid, proverDid).get();

		//7. Prover store Claim Offer received from Issuer
		Anoncreds.proverStoreClaimOffer(wallet, claimOfferJson).get();

		//8. Prover create Claim Request
		String claimReq = Anoncreds.proverCreateAndStoreClaimReq(wallet, proverDid, claimOfferJson, claimDefJson, masterSecretName).get();

		//9. Issuer open TailsReader
		JSONObject revRegDeg = new JSONObject(revRegDef);
		BlobStorage tailsReader = BlobStorage.openReader("default",
				tailsWriterConfig,
				revRegDeg.getJSONObject("value").getString("tails_location"),
				revRegDeg.getJSONObject("value").getString("tails_hash")).get();

		//10. Issuer create Claim
		int userRevocIndex = 1;
		AnoncredsResults.IssuerCreateClaimResult createClaimResult = Anoncreds.issuerCreateClaim(wallet, claimReq, gvtClaimValuesJson, revRegId, tailsReader.getTailsReaderHandle(), userRevocIndex).get();
		String claim = createClaimResult.getClaimJson();
		String revRegDelta = createClaimResult.getRevocRegDeltaJson();

		//11. Prover create RevocationInfo
		int timestamp = 100;
		String revInfo = Anoncreds.createRevocationInfo(tailsReader.getTailsReaderHandle(), revRegDef, revRegDelta, timestamp, userRevocIndex).get();

		//12. Prover store RevocationInfo
		Anoncreds.storeRevocationInfo(wallet, claimId1, revInfo).get();

		//13. Prover store received Claim
		Anoncreds.proverStoreClaim(wallet, claimId1, claim, revRegDef).get();

		//14. Prover gets Claims for Proof Request
		String claimsJson = Anoncreds.proverGetClaimsForProofReq(wallet, proofRequest).get();
		JSONObject claims = new JSONObject(claimsJson);
		JSONArray claimsForAttr1 = claims.getJSONObject("attrs").getJSONArray("attr1_referent");

		String claimUuid = claimsForAttr1.getJSONObject(0).getJSONObject("cred_info").getString("referent");

		//15. Prover create Proof
		String requestedClaimsJson = String.format("{" +
				"\"self_attested_attributes\":{}," +
				"\"requested_attrs\":{\"attr1_referent\":{\"cred_id\":\"%s\", \"revealed\":true, \"timestamp\":%d }}," +
				"\"requested_predicates\":{\"predicate1_referent\":{\"cred_id\":\"%s\", \"timestamp\":%d}}" +
				"}", claimUuid, timestamp, claimUuid, timestamp);

		String schemasJson = String.format("{\"%s\":%s}", claimUuid, schemaJson);
		String claimDefsJson = String.format("{\"%s\":%s}", claimUuid, claimDefJson);
		String revInfos = String.format("{\"%s\": { \"%s\":%s }}", claimUuid, timestamp, revInfo);

		String proofJson = Anoncreds.proverCreateProof(wallet, proofRequest, requestedClaimsJson, schemasJson, masterSecretName,
				claimDefsJson, revInfos).get();
		JSONObject proof = new JSONObject(proofJson);

		//16. Issuer revoke Claim
		revRegDelta = Anoncreds.issuerRevokeClaim(wallet, tailsReader.getTailsReaderHandle(), revRegId, userRevocIndex).get();

		//17. Verifier verify proof
		JSONObject revealedAttr1 = proof.getJSONObject("requested_proof").getJSONObject("revealed_attrs").getJSONObject("attr1_referent");
		assertEquals("Alex", revealedAttr1.getString("raw"));

		String id = revealedAttr1.getString("referent");

		schemasJson = String.format("{\"%s\":%s}", id, schemaJson);
		claimDefsJson = String.format("{\"%s\":%s}", id, claimDefJson);
		String revRegDefsJson = String.format("{\"%s\":%s}", id, revRegDef);
		String revRegs = String.format("{\"%s\": { \"%s\":%s }}", id, timestamp, revRegDelta);

		boolean valid = Anoncreds.verifierVerifyProof(proofRequest, proofJson, schemasJson, claimDefsJson, revRegDefsJson, revRegs).get();
		assertFalse(valid);

		// 17. Close and Delete Wallet
		wallet.closeWallet().get();
		Wallet.deleteWallet(walletName, null).get();
	}
}
